File: Matlab/Strategy_simple_DCA_flexible_sigma.m
function [r, num_aportaciones, valor_final_total, detalle] = Strategy_simple_DCA_flexible_sigma(periodos, basePath, tickers_config, weights_config, debug)
% Portfolio DCA simple flexible - VOL por retornos TWR y MDD por equity
% Cambios mínimos:
% - r_port (time-weighted) calculado ANTES de aportar
% - Vol anualizada desde r_port (std * sqrt(252))
% - MDD sigue desde equity_series
% - Se añade detalle.ret_series

%% ======== Validación de parámetros ========
if nargin < 3 || isempty(tickers_config)
    error('Debes especificar los tickers del portfolio');
end

if nargin < 4 || isempty(weights_config)
    weights_config = ones(1, length(tickers_config)) / length(tickers_config);
end

if nargin < 1, periodos = 5; end
if nargin < 2 || isempty(basePath), basePath = 'C:\Users\israe\OneDrive\Matlab_scripts\'; end
if nargin < 5 || isempty(debug), debug = false; end

% Validaciones
if length(tickers_config) ~= length(weights_config)
    error('Numero de tickers (%d) debe coincidir con numero de pesos (%d)', ...
        length(tickers_config), length(weights_config));
end

if abs(sum(weights_config) - 1.0) > 1e-6
    error('Los pesos deben sumar 1.0. Suma actual: %.6f', sum(weights_config));
end

%% ======== Parámetros de estrategia ========
dinero_inicial          = 1000;
aportacion_total_mes    = 1500;
dias_por_aporte         = 21;

num_activos = length(tickers_config);
w = weights_config(:);

%% ======== Verificación de archivos ========
file_paths = cell(num_activos, 1);
for i = 1:num_activos
    ticker = tickers_config{i};
    file_paths{i} = fullfile(basePath, [ticker '_yahoo.csv']);
    if ~isfile(file_paths{i})
        error('Falta archivo: %s. Ejecuta main_vYahoo.m primero.', file_paths{i});
    end
end

%% ======== Lectura de precios ========
series_data = cell(num_activos, 1);
dates_data = cell(num_activos, 1);

for i = 1:num_activos
    [series_data{i}, dates_data{i}] = read_yahoo_csv(file_paths{i});
end

%% ======== Alinear por fechas comunes ========
TT_list = cell(num_activos, 1);
for i = 1:num_activos
    TT_list{i} = timetable(dates_data{i}, series_data{i}, 'VariableNames', {tickers_config{i}});
end

if num_activos == 1
    TT = TT_list{1};
else
    TT = synchronize(TT_list{:}, 'intersection');
end
TT = rmmissing(TT);

if height(TT) == 0
    error('No quedan fechas comunes entre las series tras la interseccion.');
end

dates = TT.Properties.RowTimes;
series = table2array(TT);

%% ======== Ventana temporal ========
N_total = size(series, 1);
N_dias  = min(N_total, round(252*periodos));
series = series(end-N_dias+1:end, :);
dates = dates(end-N_dias+1:end);

%% ======== Setup para tracking completo ========
fechas_aport = 1:dias_por_aporte:N_dias;
num_aportaciones = numel(fechas_aport);
aportacion_por_activo = aportacion_total_mes * w;

% Acumulación de acciones/unidades
acc = zeros(num_activos, 1);

% Arrays de tracking
equity_series = zeros(N_dias, 1);                % equity día a día (con aportaciones)
valor_activos_series = zeros(N_dias, num_activos);
r_port = nan(N_dias,1);                           % retorno diario TWR (sin aportaciones del día)
precios_prev = series(1, :).';                    % precios día anterior para r_port

% Compra inicial (día 1) por pesos configurados
if dinero_inicial > 0
    init_por_activo = dinero_inicial * w;
    precios_d1 = series(1, :).';
    acc = acc + (init_por_activo ./ precios_d1);
end

% Tracking de aportaciones
detalle.fechas = dates(fechas_aport);
detalle.compras = zeros(num_aportaciones, num_activos);

%% ======== BUCLE PRINCIPAL CON TWR ========
aporte_idx = 1;

for dia = 1:N_dias
    precios_dia = series(dia, :).';

    % --- Retorno TWR ANTES de aportar ---
    if dia == 1
        r_port(dia) = NaN; % sin retorno el primer día
    else
        valor_inicio = sum(acc .* precios_prev); % sin flujos del día
        if valor_inicio > 0
            r_port(dia) = (sum(acc .* precios_dia) - sum(acc .* precios_prev)) / valor_inicio;
        else
            r_port(dia) = 0; % o NaN si prefieres
        end
    end

    % --- Aportación DESPUÉS de medir el retorno del día ---
    if aporte_idx <= length(fechas_aport) && dia == fechas_aport(aporte_idx)
        compra_unidades = aportacion_por_activo ./ precios_dia;
        acc = acc + compra_unidades;

        % Guardar detalles de compra
        detalle.compras(aporte_idx, :) = compra_unidades.';
        aporte_idx = aporte_idx + 1;
    end

    % Tracking de valores y equity (incluye la aportación del propio día)
    valores_dia = acc .* precios_dia;
    valor_activos_series(dia, :) = valores_dia.';
    equity_series(dia) = sum(valores_dia);

    % Actualizar precios_prev
    precios_prev = precios_dia;
end

%% ======== Valor final y métricas ========
precios_ult = series(end, :).';
valores = acc .* precios_ult;
valor_final_total = sum(valores);
pesos_finales = valores / valor_final_total;

%% ======== CAGR por IRR ========
aportes  = [dinero_inicial, repmat(aportacion_total_mes, 1, num_aportaciones)];
t_anios  = [0, (fechas_aport-1) * 365/252];
tiempos  = periodos - t_anios/365;

funcion_CAGR = @(rr) sum(aportes .* (1 + rr).^tiempos) - valor_final_total;
try
    fL = funcion_CAGR(-0.99); fR = funcion_CAGR(10);
    if ~isfinite(fL) || ~isfinite(fR)
        r = fzero(funcion_CAGR, 0.1);
    else
        r = fzero(funcion_CAGR, [-0.99, 10]);
    end
catch
    r = fzero(funcion_CAGR, 0.1);
end

%% ======== MDD (equity) y VOL (TWR) ========
mdd_val = calc_MDD_only(equity_series);
valid_r = r_port(~isnan(r_port));
daily_std = std(valid_r, 'omitnan');
annual_vol = daily_std * sqrt(252);

%% ======== Detalle estructurado con series temporales ========
acciones_struct = struct();
valores_struct = struct();
pesos_struct = struct();

for i = 1:num_activos
    ticker_clean = matlab.lang.makeValidName(tickers_config{i});
    acciones_struct.(ticker_clean) = acc(i);
    valores_struct.(ticker_clean) = valores(i);
    pesos_struct.(ticker_clean) = pesos_finales(i);
end

detalle.acciones = acciones_struct;
detalle.valor_por_activo = valores_struct;
detalle.peso_por_activo = pesos_struct;

% Series temporales y métricas
detalle.equity_series = equity_series;           % serie diaria de equity total
detalle.valor_activos_series = valor_activos_series;
detalle.dates_series = dates;
detalle.ret_series = r_port;                     % *** NUEVO: retornos diarios TWR
detalle.mdd = mdd_val;
detalle.annual_volatility = annual_vol;

detalle.params = struct('periodos', periodos, 'aportacion_total_mes', aportacion_total_mes, ...
                       'dinero_inicial', dinero_inicial, 'tickers', {tickers_config}, ...
                       'weights', weights_config, 'num_activos', num_activos);

%% ======== Salida por pantalla ========
if ~debug
    fprintf('=====================================================================================================================\n');
    fprintf('Portfolio DCA simple (%s) | %d anos | CAGR: %.2f%% | Aportes: %d | Valor final: %.2f\n', ...
            strjoin(tickers_config, '/'), periodos, r*100, num_aportaciones, valor_final_total);
    fprintf('Desglose por activo a %s:\n', datestr(dates(end), 'yyyy-mm-dd'));
    for i = 1:num_activos
        fprintf('  %s: %.2f EUR (%.2f%%)\n', tickers_config{i}, valores(i), 100*pesos_finales(i));
    end
    fprintf('  TOTAL: %.2f EUR (100%%)\n', valor_final_total);
    fprintf('=====================================================================================================================\n');
else
    fprintf('=====================================================================================================================\n');
    fprintf('Portfolio DCA Simple Flexible | Sin rebalanceo\n');
    fprintf('Tickers: %s\n', strjoin(tickers_config, ', '));
    fprintf('Pesos objetivo: %s\n', sprintf('%.1f%% ', weights_config*100));
    fprintf('Periodo: %d anos | Aportacion mensual: %.0f EUR | Inversion inicial: %.0f EUR\n', ...
            periodos, aportacion_total_mes, dinero_inicial);
    fprintf('MDD calculado: %.2f%% | Volatilidad anual (TWR): %.2f%%\n', mdd_val*100, annual_vol*100);
    fprintf('---------------------------------------------------------------------------------------------------------------------\n');
    fprintf('RESULTADOS:\n');
    fprintf('  CAGR (IRR): %.2f%%\n', r*100);
    fprintf('  Numero de aportaciones: %d\n', num_aportaciones);
    fprintf('  Valor final total: %.2f EUR\n', valor_final_total);
    fprintf('  Maximum Drawdown: %.2f%%\n', mdd_val*100);
    fprintf('  Volatilidad anualizada (TWR): %.2f%%\n', annual_vol*100);
    fprintf('---------------------------------------------------------------------------------------------------------------------\n');
    fprintf('DESGLOSE POR ACTIVO a %s:\n', datestr(dates(end), 'yyyy-mm-dd'));
    for i = 1:num_activos
        fprintf('  %s: %.2f EUR (%.2f%%) - %.4f unidades | Objetivo: %.1f%%\n', ...
            tickers_config{i}, valores(i), 100*pesos_finales(i), acc(i), weights_config(i)*100);
    end
    fprintf('  TOTAL: %.2f EUR (100%%)\n', valor_final_total);
    fprintf('=====================================================================================================================\n');
end

end

% ========================= HELPERS =========================
function [close_prices, dates] = read_yahoo_csv(csvPath)
    T = readtable(csvPath);
    if isempty(T) || width(T)==0 || height(T)==0
        error('CSV vacio o sin filas: %s', csvPath);
    end
    names = lower(strrep(strrep(T.Properties.VariableNames,'_',''),' ',''));    
    % Fecha
    iDate = find(strcmp(names,'date'),1); if isempty(iDate), iDate = 1; end
    dates = T{:,iDate};
    if ~isdatetime(dates)
        if isnumeric(dates)
            dates = datetime(dates,'ConvertFrom','datenum');
        elseif iscellstr(dates) || isstring(dates)
            try, dates = datetime(dates,'InputFormat','yyyy-MM-dd','Locale','en_US'); catch, dates = datetime(dates); end
        else
            dates = datetime(dates);
        end
    end
    % Precio (AdjClose > Close)
    iAdj   = find(strcmp(names,'adjclose'),1);
    iClose = find(strcmp(names,'close'),1);
    if ~isempty(iAdj)
        y = T{:,iAdj};
    elseif ~isempty(iClose)
        y = T{:,iClose};
    else
        y = T{:,end};
        warning('No AdjClose/Close en %s, usada ultima columna.', csvPath);
    end
    if iscellstr(y) || isstring(y), y = str2double(strrep(string(y),',','.')); end
    if ~isfloat(y), y = double(y); end
    % Limpiar y ordenar
    good = ~isnat(dates) & ~isnan(y);
    dates = dates(good); y = y(good);
    [dates, idx] = sort(dates);
    close_prices = y(idx);
    close_prices = close_prices(:); dates = dates(:);
end

function mdd = calc_MDD_only(equity_series)
    equity_series = equity_series(:);
    if numel(equity_series) < 2
        mdd = NaN;
        return;
    end
    max_run = cummax(equity_series);
    dd = (equity_series - max_run) ./ max_run;
    mdd = abs(min(dd));
end

File: Python/Strategy_simple_DCA_flexible_sigma.py
"""Portfolio DCA simple flexible - VOL por retornos TWR y MDD por equity
Cambios mínimos:
- r_port (time-weighted) calculado ANTES de aportar
- Vol anualizada desde r_port (std * sqrt(252))
- MDD sigue desde equity_series
- Se añade detalle.ret_series
"""

import os
import re
import sys

import numpy as np
import pandas as pd
from scipy.optimize import brentq, newton


def Strategy_simple_DCA_flexible_sigma(periodos=None, basePath=None, tickers_config=None,
                                       weights_config=None, debug=None):
    # ======== Validación de parámetros ========
    if _isempty(tickers_config):
        raise ValueError('Debes especificar los tickers del portfolio')

    if _isempty(weights_config):
        weights_config = np.ones(len(tickers_config)) / len(tickers_config)

    if periodos is None:
        periodos = 5
    if _isempty(basePath):
        basePath = 'C:\\Users\\israe\\OneDrive\\Matlab_scripts\\'
    if _isempty(debug):
        debug = False

    # Validaciones
    if len(tickers_config) != len(weights_config):
        raise ValueError('Numero de tickers (%d) debe coincidir con numero de pesos (%d)'
                         % (len(tickers_config), len(weights_config)))

    if abs(np.sum(weights_config) - 1.0) > 1e-6:
        raise ValueError('Los pesos deben sumar 1.0. Suma actual: %.6f' % np.sum(weights_config))

    # ======== Parámetros de estrategia ========
    dinero_inicial = 1000
    aportacion_total_mes = 1500
    dias_por_aporte = 21

    num_activos = len(tickers_config)
    w = np.asarray(weights_config, dtype=np.float64).ravel()

    # ======== Verificación de archivos ========
    file_paths = [None] * num_activos
    for i in range(1, num_activos + 1):
        ticker = tickers_config[i - 1]
        file_paths[i - 1] = os.path.join(basePath, ticker + '_yahoo.csv')
        if not os.path.isfile(file_paths[i - 1]):
            raise ValueError('Falta archivo: %s. Ejecuta main_vYahoo.m primero.' % file_paths[i - 1])

    # ======== Lectura de precios ========
    series_data = [None] * num_activos
    dates_data = [None] * num_activos

    for i in range(1, num_activos + 1):
        series_data[i - 1], dates_data[i - 1] = _read_yahoo_csv(file_paths[i - 1])

    # ======== Alinear por fechas comunes ========
    TT_list = [None] * num_activos
    for i in range(1, num_activos + 1):
        TT_list[i - 1] = pd.Series(series_data[i - 1], index=dates_data[i - 1],
                                   name=tickers_config[i - 1])

    if num_activos == 1:
        TT = TT_list[0].to_frame()
    else:
        TT = pd.concat(TT_list, axis=1, join='inner')
    TT = TT.dropna()

    if TT.shape[0] == 0:
        raise ValueError('No quedan fechas comunes entre las series tras la interseccion.')

    dates = TT.index
    series = TT.to_numpy(dtype=np.float64)

    # ======== Ventana temporal ========
    N_total = series.shape[0]
    N_dias = min(N_total, _round(252 * periodos))
    series = series[N_total - N_dias:, :]
    dates = dates[N_total - N_dias:]

    # ======== Setup para tracking completo ========
    fechas_aport = np.arange(1, N_dias + 1, dias_por_aporte)
    num_aportaciones = len(fechas_aport)
    aportacion_por_activo = aportacion_total_mes * w

    # Acumulación de acciones/unidades
    acc = np.zeros(num_activos)

    # Arrays de tracking
    equity_series = np.zeros(N_dias)                # equity día a día (con aportaciones)
    valor_activos_series = np.zeros((N_dias, num_activos))
    r_port = np.full(N_dias, np.nan)                # retorno diario TWR (sin aportaciones del día)
    precios_prev = series[0, :]                     # precios día anterior para r_port

    # Compra inicial (día 1) por pesos configurados
    if dinero_inicial > 0:
        init_por_activo = dinero_inicial * w
        precios_d1 = series[0, :]
        acc = acc + (init_por_activo / precios_d1)

    # Tracking de aportaciones
    detalle = {}
    detalle['fechas'] = dates[fechas_aport - 1]
    detalle['compras'] = np.zeros((num_aportaciones, num_activos))

    # ======== BUCLE PRINCIPAL CON TWR ========
    aporte_idx = 1

    for dia in range(1, N_dias + 1):
        precios_dia = series[dia - 1, :]

        # --- Retorno TWR ANTES de aportar ---
        if dia == 1:
            r_port[dia - 1] = np.nan  # sin retorno el primer día
        else:
            valor_inicio = np.sum(acc * precios_prev)  # sin flujos del día
            if valor_inicio > 0:
                r_port[dia - 1] = (np.sum(acc * precios_dia) - np.sum(acc * precios_prev)) / valor_inicio
            else:
                r_port[dia - 1] = 0  # o NaN si prefieres

        # --- Aportación DESPUÉS de medir el retorno del día ---
        if aporte_idx <= len(fechas_aport) and dia == fechas_aport[aporte_idx - 1]:
            compra_unidades = aportacion_por_activo / precios_dia
            acc = acc + compra_unidades

            # Guardar detalles de compra
            detalle['compras'][aporte_idx - 1, :] = compra_unidades
            aporte_idx = aporte_idx + 1

        # Tracking de valores y equity (incluye la aportación del propio día)
        valores_dia = acc * precios_dia
        valor_activos_series[dia - 1, :] = valores_dia
        equity_series[dia - 1] = np.sum(valores_dia)

        # Actualizar precios_prev
        precios_prev = precios_dia

    # ======== Valor final y métricas ========
    precios_ult = series[-1, :]
    valores = acc * precios_ult
    valor_final_total = np.sum(valores)
    pesos_finales = valores / valor_final_total

    # ======== CAGR por IRR ========
    aportes = np.concatenate(([dinero_inicial], np.full(num_aportaciones, aportacion_total_mes)))
    t_anios = np.concatenate(([0], (fechas_aport - 1) * 365 / 252))
    tiempos = periodos - t_anios / 365

    def funcion_CAGR(rr):
        return np.sum(aportes * (1 + rr) ** tiempos) - valor_final_total

    with np.errstate(invalid='ignore'):
        try:
            fL = funcion_CAGR(-0.99)
            fR = funcion_CAGR(10)
            if not np.isfinite(fL) or not np.isfinite(fR):
                r = newton(funcion_CAGR, 0.1)
            else:
                r = brentq(funcion_CAGR, -0.99, 10)
        except Exception:
            r = newton(funcion_CAGR, 0.1)

    # ======== MDD (equity) y VOL (TWR) ========
    mdd_val = _calc_MDD_only(equity_series)
    valid_r = r_port[~np.isnan(r_port)]
    daily_std = np.nanstd(valid_r, ddof=1)
    annual_vol = daily_std * np.sqrt(252)

    # ======== Detalle estructurado con series temporales ========
    acciones_struct = {}
    valores_struct = {}
    pesos_struct = {}

    for i in range(1, num_activos + 1):
        ticker_clean = _make_valid_name(tickers_config[i - 1])
        acciones_struct[ticker_clean] = acc[i - 1]
        valores_struct[ticker_clean] = valores[i - 1]
        pesos_struct[ticker_clean] = pesos_finales[i - 1]

    detalle['acciones'] = acciones_struct
    detalle['valor_por_activo'] = valores_struct
    detalle['peso_por_activo'] = pesos_struct

    # Series temporales y métricas
    detalle['equity_series'] = equity_series           # serie diaria de equity total
    detalle['valor_activos_series'] = valor_activos_series
    detalle['dates_series'] = dates
    detalle['ret_series'] = r_port                     # *** NUEVO: retornos diarios TWR
    detalle['mdd'] = mdd_val
    detalle['annual_volatility'] = annual_vol

    detalle['params'] = {'periodos': periodos, 'aportacion_total_mes': aportacion_total_mes,
                         'dinero_inicial': dinero_inicial, 'tickers': tickers_config,
                         'weights': weights_config, 'num_activos': num_activos}

    # ======== Salida por pantalla ========
    if not debug:
        print('=' * 117)
        print('Portfolio DCA simple (%s) | %d anos | CAGR: %s%% | Aportes: %d | Valor final: %s'
              % ('/'.join(tickers_config), int(periodos), _fmt(r * 100, '.2f'),
                 num_aportaciones, _fmt(valor_final_total, '.2f')))
        print('Desglose por activo a %s:' % pd.Timestamp(dates[-1]).strftime('%Y-%m-%d'))
        for i in range(1, num_activos + 1):
            print('  %s: %s EUR (%s%%)'
                  % (tickers_config[i - 1], _fmt(valores[i - 1], '.2f'),
                     _fmt(100 * pesos_finales[i - 1], '.2f')))
        print('  TOTAL: %s EUR (100%%)' % _fmt(valor_final_total, '.2f'))
        print('=' * 117)
    else:
        print('=' * 117)
        print('Portfolio DCA Simple Flexible | Sin rebalanceo')
        print('Tickers: %s' % ', '.join(tickers_config))
        print('Pesos objetivo: %s' % ''.join('%.1f%% ' % (x * 100) for x in np.asarray(weights_config).ravel()))
        print('Periodo: %d anos | Aportacion mensual: %s EUR | Inversion inicial: %s EUR'
              % (int(periodos), _fmt(aportacion_total_mes, '.0f'), _fmt(dinero_inicial, '.0f')))
        print('MDD calculado: %s%% | Volatilidad anual (TWR): %s%%'
              % (_fmt(mdd_val * 100, '.2f'), _fmt(annual_vol * 100, '.2f')))
        print('-' * 117)
        print('RESULTADOS:')
        print('  CAGR (IRR): %s%%' % _fmt(r * 100, '.2f'))
        print('  Numero de aportaciones: %d' % num_aportaciones)
        print('  Valor final total: %s EUR' % _fmt(valor_final_total, '.2f'))
        print('  Maximum Drawdown: %s%%' % _fmt(mdd_val * 100, '.2f'))
        print('  Volatilidad anualizada (TWR): %s%%' % _fmt(annual_vol * 100, '.2f'))
        print('-' * 117)
        print('DESGLOSE POR ACTIVO a %s:' % pd.Timestamp(dates[-1]).strftime('%Y-%m-%d'))
        for i in range(1, num_activos + 1):
            print('  %s: %s EUR (%s%%) - %s unidades | Objetivo: %s%%'
                  % (tickers_config[i - 1], _fmt(valores[i - 1], '.2f'),
                     _fmt(100 * pesos_finales[i - 1], '.2f'), _fmt(acc[i - 1], '.4f'),
                     _fmt(np.asarray(weights_config).ravel()[i - 1] * 100, '.1f')))
        print('  TOTAL: %s EUR (100%%)' % _fmt(valor_final_total, '.2f'))
        print('=' * 117)

    return r, num_aportaciones, valor_final_total, detalle


# ========================= HELPERS =========================
def _read_yahoo_csv(csvPath):
    T = pd.read_csv(csvPath)
    if T.empty or T.shape[1] == 0 or T.shape[0] == 0:
        raise ValueError('CSV vacio o sin filas: %s' % csvPath)
    names = [c.replace('_', '').replace(' ', '').lower() for c in T.columns]
    # Fecha
    iDate = next((i for i, n in enumerate(names) if n == 'date'), None)
    if iDate is None:
        iDate = 0
    dates = T.iloc[:, iDate]
    if not pd.api.types.is_datetime64_any_dtype(dates):
        if pd.api.types.is_numeric_dtype(dates):
            dates = pd.to_datetime(dates, unit='D', origin='1899-12-30')
        elif pd.api.types.is_object_dtype(dates) or pd.api.types.is_string_dtype(dates):
            try:
                dates = pd.to_datetime(dates, format='%Y-%m-%d')
            except (ValueError, TypeError):
                dates = pd.to_datetime(dates)
        else:
            dates = pd.to_datetime(dates)
    # Precio (AdjClose > Close)
    iAdj = next((i for i, n in enumerate(names) if n == 'adjclose'), None)
    iClose = next((i for i, n in enumerate(names) if n == 'close'), None)
    if iAdj is not None:
        y = T.iloc[:, iAdj]
    elif iClose is not None:
        y = T.iloc[:, iClose]
    else:
        y = T.iloc[:, -1]
        print('Warning: No AdjClose/Close en %s, usada ultima columna.' % csvPath, file=sys.stderr)
    if pd.api.types.is_object_dtype(y) or pd.api.types.is_string_dtype(y):
        y = pd.to_numeric(
            pd.Series(y).astype(str).str.replace(',', '.', regex=False),
            errors='coerce')
    y = np.asarray(y, dtype=np.float64)
    # Limpiar y ordenar
    dates = pd.DatetimeIndex(dates)
    good = dates.notna() & ~np.isnan(y)
    dates = dates[good]
    y = y[good]
    idx = np.argsort(dates, kind='stable')
    dates = dates[idx]
    close_prices = y[idx]
    close_prices = np.asarray(close_prices, dtype=np.float64).ravel()
    dates = pd.DatetimeIndex(dates)
    return close_prices, dates


def _isempty(x):
    if x is None:
        return True
    if isinstance(x, (str, bytes, list, tuple, dict)):
        return len(x) == 0
    if isinstance(x, np.ndarray):
        return x.size == 0
    return False


def _round(x):
    return int(np.sign(x) * np.floor(np.abs(x) + 0.5))


def _make_valid_name(s):
    s = re.sub(r'\W', '_', s)
    if s and s[0].isdigit():
        s = 'x' + s
    return s


def _fmt(x, code):
    if isinstance(x, (float, np.floating)):
        if np.isnan(x):
            return 'NaN'
        if np.isinf(x):
            return 'Inf' if x > 0 else '-Inf'
    return format(x, code)


def _calc_MDD_only(equity_series):
    equity_series = np.asarray(equity_series, dtype=np.float64).ravel()
    if equity_series.size < 2:
        mdd = np.nan
        return mdd
    max_run = np.maximum.accumulate(equity_series)
    dd = (equity_series - max_run) / max_run
    mdd = abs(np.min(dd))
    return mdd
